src/smefit/optimize/ultranest.py

```diff
@@ -190,44 +207,106 @@ def from_dict(cls, config):
             target_post_unc=target_post_unc,
             frac_remain=frac_remain,
             store_raw=store_raw,
+            vectorized=vectorized,
             external_chi2=external_chi2,
         )
 
     def chi2_func_ns(self, params):
-        """Wrap the chi2 in a function for the optimizer. Pass noise and
-        data info as args. Log the chi2 value and values of the coefficients.
+        """Compute the chi2 function for |NS|.
+        It is simplified with respect to the one in the Optimizer class,
+        so that it can be compiled with jax.jit.
+        Parameters
+        ----------
+        params : jnp.ndarray
+            Wilson coefficients
+        """
+
+        if self.loaded_datasets is not None:
+            chi2_tot = chi2.compute_chi2(
+                self.loaded_datasets,
+                params,
+                self.use_quad,
+                self.use_multiplicative_prescription,
+                use_replica=False,
+            )
+        else:
+            chi2_tot = 0
+
+        if self.chi2_ext is not None:
+            for chi2_ext in self.chi2_ext:
+                chi2_ext_i = chi2_ext(params)
+                chi2_tot += chi2_ext_i
+
+        return chi2_tot
+
+    def compute_fixed_coeff(self, constrain, param_dict):
+        """Compute the fixed coefficient."""
+        temp = 0.0
+        for add_factor_dict in constrain:
+            free_dofs = jnp.array(
+                [param_dict[fixed_name] for fixed_name in add_factor_dict]
+            )
+            fact_exp = jnp.array(list(add_factor_dict.values()), dtype=float)
+            temp += jnp.prod(fact_exp[:, 0] * jnp.power(free_dofs, fact_exp[:, 1]))
+        return temp
+
+    def produce_all_params(self, params):
+        """Produce all parameters from the free parameters.
 
         Parameters
         ----------
-        params : np.ndarray
-            noise and data info
+        params : jnp.ndarray
+            free parameters
 
         Returns
         -------
-        current_chi2 : np.ndarray
-            chi2 function
+        all_params : jnp.ndarray
+            all parameters
         """
-        self.coefficients.set_free_parameters(params)
-        self.coefficients.set_constraints()
+        is_free = self.coefficients.is_free
```

Comment:
instead of adding here something similar to `set_constraints()`, can't  you just adapt the code there? 
This way I see some steps end up duplicated, see for instance `compute_fixed_coeff`

Reply:
Yes, there is duplication. The problem is that I struggled to use those already implemented methods, because the are "object-oriented" and JAX works instead in a functional paradigm with pure functions... So basically when I was trying to use self.set_constraints(), it was not able to jit compile them.

But I understand it's not very elegant, so open to other ideas to make it work

Reply:
I see maybe I can have a try to solve this issue.